src/adapter/src/peek_client.rs

```diff
@@ -0,0 +1,381 @@
+    /// we intentionally do not special‑case sinks here because peeks never read from sinks.
+    /// Therefore, using `StorageCollections::collections_frontiers` is sufficient.
+    ///
+    /// Note: self is taken &mut because of the lazy fetching in `get_compute_instance_client`.
+    pub async fn acquire_read_holds_and_collection_write_frontiers(
+        &mut self,
+        id_bundle: &CollectionIdBundle,
+    ) -> Result<(ReadHolds<Timestamp>, Antichain<Timestamp>), CollectionMissing> {
+        let mut read_holds = ReadHolds::new();
+        let mut upper = Antichain::new();
+
+        if !id_bundle.storage_ids.is_empty() {
+            let desired_storage: Vec<_> = id_bundle.storage_ids.iter().copied().collect();
+            let storage_read_holds = self
+                .storage_collections
+                .acquire_read_holds(desired_storage)
+                .expect("missing storage collections");
+            read_holds.storage_holds = storage_read_holds
+                .into_iter()
+                .map(|hold| (hold.id(), hold))
+                .collect();
+
+            let storage_ids: Vec<_> = id_bundle.storage_ids.iter().copied().collect();
+            for f in self
+                .storage_collections
+                .collections_frontiers(storage_ids)
+                .expect("missing collections")
+            {
+                upper.extend(f.write_frontier);
+            }
+        }
+
+        for (&instance_id, collection_ids) in &id_bundle.compute_ids {
+            let client = self
+                .ensure_compute_instance_client(instance_id)
+                .await
+                .expect("missing compute instance client");
+
+            for (id, read_hold, write_frontier) in client
+                .acquire_read_holds_and_collection_write_frontiers(
+                    collection_ids.iter().copied().collect(),
+                )
+                .await?
+            {
+                let prev = read_holds
+                    .compute_holds
+                    .insert((instance_id, id), read_hold);
+                assert!(
+                    prev.is_none(),
+                    "duplicate compute ID in id_bundle {id_bundle:?}"
+                );
+
+                upper.extend(write_frontier);
+            }
+        }
+
+        Ok((read_holds, upper))
+    }
+
+    /// Implement a fast-path peek plan.
+    /// This is similar to `Coordinator::implement_peek_plan`, but only for fast path peeks.
+    ///
+    /// Supported variants:
+    /// - FastPathPlan::Constant
+    /// - FastPathPlan::PeekExisting (PeekTarget::Index only)
+    ///
+    /// This fn assumes that the caller has already acquired read holds for the peek at the
+    /// appropriate timestamp. Before this fn returns, it passes on the responsibility of holding
+    /// back the frontiers to Compute, so then the caller can forget its own read holds.
```

Comment:
I think this is outdated now that we pass in the input read holds explicitly?

Reply:
Yes, sorry!